processing/src/main/java/org/apache/druid/query/scan/OrderByLimitSequence.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.scan;
+
+import com.google.common.collect.Lists;
+import org.apache.druid.java.util.common.DateTimes;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.UOE;
+import org.apache.druid.java.util.common.guava.BaseSequence;
+import org.apache.druid.query.QueryTimeoutException;
+import org.apache.druid.query.context.ResponseContext;
+import org.apache.druid.segment.BaseObjectColumnValueSelector;
+import org.apache.druid.segment.Cursor;
+import org.apache.druid.segment.column.ColumnHolder;
+import org.apache.druid.timeline.SegmentId;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+class OrderByLimitSequence extends BaseSequence<ScanResultValue, Iterator<ScanResultValue>>
```

Comment:
Once you have the basic sort/merge working, then you can add limits. There are two cases.

In the sort phase, it is necessary to sort the entire result set to apply the limit after sorting. So, you can simply plop a limit sequence on top of your order sequence, which is what we do in the operator case.

Then, in the merge, you can just have one sequence do the merge, ignorant of a limit. Have a limit sequence sit on top. The limit sequence just stops reading from the merge, and closes it, when it has reached the limit.

In this way, each operation is separate and simple. You don't need to have the sort and merge also know how to do the limit.

It would be the query runner that works out how to stack these sequences. We'd have to work out exactly which one does the deed, but my guess is that it is the one that works with the scan engine: it just inserts extra sequences as needed for sort/merge/limit.